### PR TITLE
PXC-3990: Building galera using CMake

### PR DESCRIPTION
**Pull request by dutow**
This commit modifies the galera source so it builds correctly as a
subproject of PXC, with MYSQL_MAINTAINER_MODE=ON

Source changes:
 * Changed fallthrough hints
 * Added __attribute__((format)) where required
 * Changed the signature of gu_log to support the format attribute
 * Fixed incorrect format strings detected by the above
 * Disabled variable hint array warnings
 * Modified some C-style casts to keep constness where possible
 * Modified them to const/reinterpret cast otherwise
 * Unnecessary ; are removed
 * Unnecessary GNU macros (e.g. _GNU_SOURCE) removed
 * EAGAIN and EWOULDBLOCK is the same, only kept one with a static
   assert
 * Corrected some #if to #ifdef
 * Galerautils tests are built without -Wformat because of use of
   deprecated check functions
 * GCS_FOR_GARB definition should be public for gcs4garb target

### CODE REVIEW

File: galera/src/certification.cpp
Comment (venkatesh-prasad-v) [inline galera/src/certification.cpp:188]: Here and everywhere. Wouldn't it be better to use C++17 `[[fallthrough]]` attribute? 

Currently, in all other places we are using [[fallthrough]] attribute since https://github.com/percona/percona-server/commit/c5923aa1aedea4b4268e2b372098e2bf

Comment (dutow) [reply galera/src/certification.cpp:188]: Galera sets CMAKE_CXX_STANDARD to 11.

**Comment by venkatesh-prasad-v** [comment]
Hi! I found one more issue with cmake. Seems that the peformance_schema instrumentation in galera doesn't work with the current cmake changes.
```
WITH scons psi=1
----------------

mysql> select * from cond_instances where name like '%galera%';
+---------------------------------------------------------+-----------------------+
| NAME                                                    | OBJECT_INSTANCE_BEGIN |
+---------------------------------------------------------+-----------------------+
| wait/synch/cond/galera/COND_galera_sst                  |        94453843770064 |
| wait/synch/cond/galera/COND_galera_service_thd          |        94453843778832 |
| wait/synch/cond/galera/COND_galera_service_thd_flush    |        94453843778928 |
| wait/synch/cond/galera/COND_galera_ist_receiver         |        94453843785824 |
| wait/synch/cond/galera/COND_galera_async_sender_monitor |        94453843779120 |
| wait/synch/cond/galera/COND_galera_local_monitor        |        94453843969776 |
| wait/synch/cond/galera/COND_galera_apply_monitor        |        94453843969968 |
| wait/synch/cond/galera/COND_galera_commit_monitor       |        94453843970160 |
| wait/synch/cond/galera/COND_galera_recvbuf              |        94453844000448 |
+---------------------------------------------------------+-----------------------+
9 rows in set (0.00 sec)

mysql> select * from file_instances where file_name like '%galera%';
+---------------------------------------------------+--------------------------------------------+------------+
| FILE_NAME                                         | EVENT_NAME                                 | OPEN_COUNT |
+---------------------------------------------------+--------------------------------------------+------------+
| /dev/shm/var_auto_fcmq/mysqld.1/data/galera.cache | wait/io/file/galera/FILE_galera_ringbuffer |          1 |
+---------------------------------------------------+--------------------------------------------+------------+
1 row in set (0.00 sec)

WITH CMAKE
----------
mysql> select * from file_instances where file_name like '%galera%';
Empty set (0.00 sec)

mysql> select * from cond_instances where name like '%galera%';
Empty set (0.00 sec)
```

**Review by venkatesh-prasad-v — APPROVED** [review]
Approved. Please wait for the final approval from Kamil. Also, please squash all commits before merging.

File: galerautils/src/gu_asio_stream_engine.cpp
Comment (kamil-holubicki) [inline galerautils/src/gu_asio_stream_engine.cpp:79]: Why this change? Are they always the same?

Comment (dutow) [reply galerautils/src/gu_asio_stream_engine.cpp:79]: Because they are the same, and generate warnings with the warning settings in the server code. I've added static_asserts to ensure that they are indeed the same on all platforms where we build pxc.

Comment (kamil-holubicki) [reply galerautils/src/gu_asio_stream_engine.cpp:79]: I mean if it is guaranteed that they are the same, I'm fine with this, but I see other places in the code (not Galera), where there is a distinction between the two.
For example viosocket.cc solves this problem is this way:
```
#if SOCKET_EAGAIN == SOCKET_EWOULDBLOCK
    if (error != SOCKET_EAGAIN)
#else
    if (error != SOCKET_EAGAIN && error != SOCKET_EWOULDBLOCK)
#endif
```

Comment (dutow) [reply galerautils/src/gu_asio_stream_engine.cpp:79]: Yes, that's why I also added a static assert. I did a search about this change, and found this:
https://stackoverflow.com/questions/7003234/which-systems-define-eagain-and-ewouldblock-as-different-values/7003379#7003379

Which explains that they are only different on some old unix systems (SysV, HPUX, old AIX version, OS/390).

I doubt we'll ever build pxc on any of those.

**Review by kamil-holubicki — APPROVED** [review]

**Pull request merged**
